File: Stack/stack.py
class Stack:
    def __init__(self):
        self.cap=4
        self.count=0
        self.data= [None]*self.cap
        
    def resize(self):
        self.cap*=2
        newData= [None]*self.cap
        for i in range (self.count):
            newData[i]= self.data[i]
            
        self.data= newData
        
    def push(self, val):
        if self.cap==self.count:
            self.resize()
        self.data[self.count]= val
        self.count+=1
        
    def pop(self):
        if self.empty():
            raise IndexError("Empty Stack")
        
        val= self.data[self.count-1]
        self.data[self.count-1]= None
        
        return val
    
    def top(self):
        if self.empty():
            raise IndexError("Empty Stack")
        
        return self.data[self.count-1]
    
    def empty(self):
        return (self.count==0)

File: Stack/stack.cpp
template <typename T>
class Stack
{
    private:
        T* data;
        int cap;
        int count;
    
        void resize()
        {
            int cap= cap*2;
            T* newData= new T[cap];

            for (int i=0; i<count; i++)
            {
                newData[i]= data[i];
            }

            delete[] data;
            data= newData;
        }
    public:
        Stack()
        {
            cap= 4;
            count=0;
            data= new T[cap];
        }

        ~Stack()
        {
            delete[] data;
        }

        void push(const T& val)
        {
            if (count==cap)
            {
                resize();
            }
            data[count]= value;
            count++;
        }

        void pop()
        {
            if (count==0)
            {
                throw std::runtime_error("Empty Stack");
            }
            count--;
        }

        T& top()
        {
            if (count==0)
            {
                throw std::runtime_error("Empty Stack");
            } 
            return data[count-1];
        }

        bool empty()
        {
            return (count==0);
        }

};
